Restart the init container to not be stuck in created state

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind regression
/kind bug

#### What this PR does / why we need it:
This restarts the init container if it is stuck in the created state, as it may have failed to start correctly.
This fixes a regression from https://github.com/kubernetes/kubernetes/pull/116429.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #126440

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a regression in default 1.29 configurations with the SidecarContainers feature enabled, where init containers may fail to start due to a temporary container runtime failure.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

gjkim42 committed:
Restart the init container to not be stuck in created state

The main sync loop should have created and started the container in one
step. If the init container is in the 'created' state, it's likely that
the container runtime failed to start it. To prevent the container from
getting stuck in the 'created' state, restart it.

gjkim42 commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

toVersus commented on pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:
```diff
@@ -1170,6 +1170,17 @@ func TestComputePodActions(t *testing.T) {
 				ContainersToStart: []int{1},
 			},
 		},
+		"Restart the container if the container is in created state": {
```
I see, you're adding a test case here because the kubelet handles the Regular Containers [the same way](https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/pkg/kubelet/container/helpers.go#L89-L92).

toVersus reviewed (commented):
The change looks good to me, but you're planning to include the E2E tests in this PR if we get the scenario that causes the issue and can turn it into a test, right?

gjkim42 commented:
> The change looks good to me, but you're planning to include the E2E tests in this PR if we get the scenario that causes the issue and can turn it into a test, right?

I intended to do that, but since it is difficult to simulate the temporary container runtime failure, fixing it with unit tests should be ok.

bart0sh commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -1084,7 +1084,8 @@ func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, pod
 		case kubecontainer.ContainerStateCreated:
-			// nothing to do but wait for it to start
+			// restart the container if it is stuck in the created state
+			changes.InitContainersToStart = append(changes.InitContainersToStart, i)
```
Can you explain why container considered stuck if it's in the `ContainerStateCreated` state?

gjkim42 replied:
That is because of the old code that does nothing in the Created state.
I'll rephrase it.

bart0sh replied:
It still looks to me like we're trying to unconditionally restart container if it's in the `ContainerStateCreated` state. Where this state is set? Where is it updated? Why is it `likely that the container runtime has failed to start the container` if the container is in this state?

/hold

gjkim42 replied:
This is a part of the main sync loop and
1. it checks the state
2. calls the CreateContainer
3. calls the StartContainer

in sequence with blocking manner (waiting for the result).

Therefore, if a container is in the Created state at step 1, it means that the container runtime has failed to StartContainer.

https://github.com/kubernetes/kubernetes/pull/126543/files#r1704785057
^
The regular containers behave in the same way.

bart0sh replied:
It does. Thanks for the explanations!
/unhold

hshiina replied:
This can cause the same issue as #123087 for init containers when `EventedPLEG` is enabled.

Container statuses evaluated here are updated asynchronously by PLEG. So, `ContainerStateCreated` could be seen when a container is starting normally especially with `EventedPLEG` enabled.
xref: https://github.com/kubernetes/kubernetes/pull/122778#issuecomment-1956910606

toVersus replied:
Thanks for sharing! I see that the change to prevent Regular containers from restarting in the Created state when Evented PLEG is enabled will be made in https://github.com/kubernetes/kubernetes/pull/122778.

This PR fixes a regression in Generic PLEG caused by Sidecar Containers feature. We could either merge this PR first and then address the case for initContainers in https://github.com/kubernetes/kubernetes/pull/122778, or we could take a different approach. Based on the  https://github.com/kubernetes/enhancements/issues/3386#issuecomment-2309247626, it looks like there might be some progress with https://github.com/kubernetes/kubernetes/pull/122778 in near future too.

/cc @pacoxu

pacoxu replied:
This addressed the problem of Generic PLEG. 

/lgtm
/approve

ffromani replied:
@gjkim42 if (and only if) you happen to resubmit, could you please add more of your comment https://github.com/kubernetes/kubernetes/pull/126543/commits/7e8ea57cc6ee29631e8a66cc521a48803a16cd14#r1724882374 in the code comment? I feel the github comment concisely and clearly describes the fix so it would be great to have it in the codebase and not lost in github review comments.

gjkim42 replied:
resubmitted.

kannon92 commented:
/triage accepted

kannon92 commented:
/cc @SergeyKanzhelev

matthyx commented:
/lgtm

pacoxu commented:
/assign @mrunalp @yujuhong @smarterclayton

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -1084,7 +1084,10 @@ func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, pod
 
```
computeInitContainerActions looks like a new code path only enabled when the SidecarContainers feature is enabled?

taking a step back, should we only take this new code path in pods which actually have made use of the sidecar feature? ideally, bugs in this new code path would only impact pods that have specifically opted into using the new sidecar feature, and all other pods would have their behavior stay the same.

gjkim42 replied:
I would like to remove the old code path while updating the new one once the SidecarContainers feature reaches GA, since there is a lot of shared logic for the init container's actions.

liggitt replied:
> I would like to remove the old code path while updating the new one once the SidecarContainers feature reaches GA, since there is a lot of shared logic for the init container's actions.

The number one priority is to ensure that pods that didn't opt into the new feature are not impacted by bugs in the new code path. This is the second time the sidecar feature has regressed, so I think it's important to leave the old code path active for pods which have not opted in until the new code path has proven it works correctly over multiple releases.

gjkim42 replied:
I agree that it should have been implemented in that way, but it might be a bit late now since the sidecar wg is considering to promote this feature to GA in the next release...

cc @SergeyKanzhelev

liggitt replied:
> it might be a bit late now since the sidecar wg is considering to promote this feature to GA in the next release...

I do not agree. The presence of a regression in this code path should make us hesitant to make this code path unconditional. Recovering compatibility for 1.29-1.31 should be our top priority. Reviewing these changes and closing test gaps should be our second priority. Promotion to GA should come after all of those.

gjkim42 replied:
ok, I'll create a PR for that to ensure that the minimize the impact of the new code path.

ffromani commented:
/lgtm

only changed comment from the last LGTM

gjkim42 commented:
/assign @SergeyKanzhelev @smarterclayton 

Could you please take a look? 
This will resolve the regression where the init container could get stuck in the 'created' state due to a partial failure (successfully created but failed to start).

gjkim42 commented:
/retest

gjkim42 commented:
https://github.com/kubernetes/kubernetes/pull/127162
made a PR to fix the regression related to the regular init containers first.

still need this PR to fix the bug where restartable init containers can be stuck in the created state.

gjkim42 commented:
manually tested this PR with customized containerd (that fails to start container at first after 10s) and confirmed that this PR fixed this bug with regular and restartable init containers.

```
NAME                                   READY   STATUS     RESTARTS   AGE                                                                                                                            [3/1683]
fixed-k8s-regular-init-container       0/1     Init:0/1   0          7s               
fixed-k8s-restartable-init-container   0/2     Init:0/1   0          7s               
k8s-1.30-regular-init-container        0/1     Init:0/1   0          7s               
k8s-1.30-restartable-init-container    0/2     Init:0/1   0          7s               
NAME                                   READY   STATUS     RESTARTS   AGE              
fixed-k8s-regular-init-container       0/1     Init:0/1   0          9s               
fixed-k8s-restartable-init-container   0/2     Init:0/1   0          9s               
k8s-1.30-regular-init-container        0/1     Init:0/1   0          9s               
k8s-1.30-restartable-init-container    0/2     Init:0/1   0          9s               
NAME                                   READY   STATUS     RESTARTS   AGE              
fixed-k8s-regular-init-container       0/1     Init:0/1   0          12s              
fixed-k8s-restartable-init-container   0/2     Init:0/1   0          12s              
k8s-1.30-regular-init-container        0/1     Init:0/1   0          12s              
k8s-1.30-restartable-init-container    0/2     Init:0/1   0          12s              
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       0/1     Init:RunContainerError   0          14s
fixed-k8s-restartable-init-container   0/2     Init:0/1                 0          14s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          14s
k8s-1.30-restartable-init-container    0/2     Init:0/1                 0          14s
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       0/1     Init:0/1                 1          16s
fixed-k8s-restartable-init-container   0/2     Init:RunContainerError   0          16s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          16s
k8s-1.30-restartable-init-container    0/2     Init:RunContainerError   0          16s
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       0/1     Init:0/1                 1          18s
fixed-k8s-restartable-init-container   1/2     PodInitializing          1          18s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          18s                                                                                                                      
k8s-1.30-restartable-init-container    0/2     Init:RunContainerError   0          18s
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       0/1     Init:0/1                 1          20s
fixed-k8s-restartable-init-container   2/2     Running                  1          20s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          20s
k8s-1.30-restartable-init-container    0/2     Init:RunContainerError   0          20s
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       0/1     Init:0/1                 1          22s
fixed-k8s-restartable-init-container   2/2     Running                  1          22s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          22s
k8s-1.30-restartable-init-container    0/2     Init:RunContainerError   0          22s
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       0/1     Init:0/1                 1          24s
fixed-k8s-restartable-init-container   2/2     Running                  1          24s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          24s
k8s-1.30-restartable-init-container    0/2     Init:RunContainerError   0          24s
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       0/1     PodInitializing          0          26s
fixed-k8s-restartable-init-container   2/2     Running                  1          26s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          26s
k8s-1.30-restartable-init-container    0/2     Init:RunContainerError   0          26s
NAME                                   READY   STATUS                   RESTARTS   AGE
fixed-k8s-regular-init-container       1/1     Running                  0          28s
fixed-k8s-restartable-init-container   2/2     Running                  1          28s
k8s-1.30-regular-init-container        0/1     Init:RunContainerError   0          28s
k8s-1.30-restartable-init-container    0/2     Init:RunContainerError   0          28s
```

```
$ k get nodes
NAME                 STATUS   ROLES           AGE     VERSION
kind-control-plane   Ready    control-plane   2m44s   v1.32.0-alpha.0.171+4363a8453cef12-dirty
kind-worker          Ready    <none>          2m33s   v1.30.4
kind-worker2         Ready    <none>          2m32s   v1.32.0-alpha.0.171+4363a8453cef12-dirty
```

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

this matches the behavior of a regular containers: https://github.com/kubernetes/kubernetes/blob/d28fca12944f12a778392147b862c24c48289544/pkg/kubelet/container/helpers.go#L89-L92

It was an interesting code digging to understand how it works. Thank you @gjkim42!
